go/src/koding/klient/machine/mount/sync/history/history.go

```diff
@@ -0,0 +1,161 @@
+package history
+
+import (
+	"container/ring"
+	"fmt"
+	"sync"
+	"time"
+
+	msync "koding/klient/machine/mount/sync"
+)
+
+// Record stores a single history record.
+type Record struct {
+	CreatedAt time.Time `json:"createdAt"`        // creation time.
+	Message   string    `json:"message"`          // short summary.
+	Details   string    `json:"detail,omitempty"` // detailed description.
+}
+
+// History gathers synchronization history of results produced by stored Syncer.
+type History struct {
+	s msync.Syncer // underlying Syncer.
+
+	mu sync.Mutex
+	r  *ring.Ring
+
+	once  sync.Once
+	stopC chan struct{} // channel used to close any opened exec streams.
+}
+
+// NewHistory creates a new History instance. Provided size defines the maximum
+// length of history records.
+func NewHistory(s msync.Syncer, size int) *History {
+	return &History{
+		s:     s,
+		r:     ring.New(size),
+		stopC: make(chan struct{}),
+	}
+}
+
+// ExecStream wraps underlying execers with history gathering logic.
+func (h *History) ExecStream(evC <-chan *msync.Event) <-chan msync.Execer {
+	exhC := make(chan msync.Execer)
+
+	go func() {
+		defer close(exhC)
+
+		exC := h.s.ExecStream(evC)
+		for {
+			select {
+			case ex, ok := <-exC:
+				if !ok {
+					return
+				}
+
+				h.add(&Record{
+					CreatedAt: time.Now().UTC(),
+					Message:   "received: " + ex.String(),
+				})
+
+				exh := &histExec{
+					ex:     ex,
+					parent: h,
+				}
+
+				select {
+				case exhC <- exh:
+				case <-h.stopC:
+					return
+				}
+			case <-h.stopC:
+				return
+			}
+		}
+	}()
+
+	return exhC
+}
+
+// Close stops all created synchronization streams.
+func (h *History) Close() error {
+	h.once.Do(func() {
+		close(h.stopC)
+	})
+
+	return nil
+}
+
+// Get gets recorded history. Returned slice length will not exceed history
+// maximum size.
+func (h *History) Get() []*Record {
+	recs := make([]*Record, 0)
```

Comment:
> What's tricky here?

Formatting of `kd machine mount inspect` output should not rely on internal api (empty slice vs nil slice), especially when the behaviour is undocumented. The command handler itself should inspect if there's no history, and e.g. display `No history available.` or alike.

In some future some other developer can spot opportunity for refactoring, changing the line according to my suggestion, not knowing that it could break some other code elsewhere, especially when there's no test for this.